### PR TITLE
ADBDEV-4902-9: Fix null checks in `cdbpullup_findEclassInTargetList()`

### PR DESCRIPTION
**Pull request by bandetto**
Fix null checks in cdbpullup_findEclassInTargetList()

Removed extra while loop. All values of key with RelabelType kind will be
already skipped at this point by previous while loop with the same condition.
There is no point in the second one.

Added null checks in dangerous places. There is tlist_member_ignore_relabel()
and tlist_member_match_var() which have similar null checks.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/63347

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1035953

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1035954

File: src/backend/cdb/cdbpullup.c [hunk 1]
```diff
@@ -360,16 +360,16 @@ cdbpullup_findEclassInTargetList(EquivalenceClass *eclass, List *targetlist,
 			 * Check if targetlist is a List of TargetEntry. (Planner's
 			 * RelOptInfo targetlists don't have TargetEntry nodes.)
 			 */
-			if (IsA(tlexpr, TargetEntry))
+			if (tlexpr && IsA(tlexpr, TargetEntry))
```
Comment (RekGRpth) [inline src/backend/cdb/cdbpullup.c]: Why may tlexpr be NULL here?

File: src/backend/cdb/cdbpullup.c [hunk 2]
```diff
@@ -382,17 +382,14 @@ cdbpullup_findEclassInTargetList(EquivalenceClass *eclass, List *targetlist,
 			}
 			else
 			{
-				/* ignore RelabelType nodes on both sides */
-				while (key && IsA(key, RelabelType))
-					key = (Expr *) ((RelabelType *) key)->arg;
-
-				if (equal(tlexpr, key))
+				if ((tlexpr || key) && equal(tlexpr, key))
```
Comment (RekGRpth) [inline src/backend/cdb/cdbpullup.c]: Why did you add additional conditions?

https://github.com/arenadata/gpdb/blob/3b5aa3e9d5141832d94172e20e41a75770c07f34/src/backend/nodes/equalfuncs.c#L2843-L2849

Comment (bandetto) [reply src/backend/cdb/cdbpullup.c]: So that the behavior is not changed. Without this check, if `key = NULL` and `tlexpr = NULL`, `NULL` will be returned without processing remaining values from targetlist.

Comment (RekGRpth) [reply src/backend/cdb/cdbpullup.c]: if 

> behavior is not changed



> Why did you add additional conditions?

**Comment by RekGRpth** [comment]
update base branch to 4726

File: src/backend/cdb/cdbpullup.c (hunk 2 above)
Comment (bandetto) [reply src/backend/cdb/cdbpullup.c]: If `while (key && IsA(key, RelabelType))` stops because `key == NULL`, and then `while (tlexpr && IsA(tlexpr, RelabelType))` stops because `tlexpr == NULL`, we will get into `else` branch, where `equal(tlexpr, key)` will return `true` and the function will return `NULL`. Is this behavior desired? With this check, current `key` will be skipped, and loop will continue. If the none of the keys match, we will return `NULL` anyway.

Right now this case might not happen. If both keys somehow happen to be `NULL`, this condition will stop us exiting from the loop too soon. 

These values were assumed to be non-`NULL` anyway. Should `if ((tlexpr || key) && equal(tlexpr, key))` be changed to just `equal(tlexpr, key))`?

**Comment by BenderArenadata** [comment]
Allure report https://allure-ee.adsw.io/launch/63445

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1040835

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1040836

**Comment by RekGRpth** [comment]
> There is tlist_member_ignore_relabel()
> and tlist_member_match_var() which have similar null checks.

How does this relate to the changes in the current patch? If those functions have the same bug, can they also be fixed in this patch?

**Comment by bandetto** [comment]
> How does this relate to the changes in the current patch? If those functions have the same bug, can they also be fixed in this patch?

They are search functions too, and they are checking each entry in a targetlist similarly. They already had `NULL` checks, but `cdbpullup_findEclassInTargetList()` didn't.

File: src/backend/cdb/cdbpullup.c (hunk 2 above)
Comment (RekGRpth) [reply src/backend/cdb/cdbpullup.c]: Is this a refactoring, not a bug fix?

**Review by RekGRpth — APPROVED** [review]
